src/utils/BucketUtils.js

```diff
@@ -0,0 +1,251 @@
+  const flatTranslatedDataUses = map(t => compact(t))(translatedDataUses);
+
+  // Step 1: Create buckets for unique dataset groups
+  map(d => {
+    // Put each dataset into a bucket. If the dataset's data use is missing, unique or has an "Other" restriction, then
+    // it gets its own bucket. If the data use is already in a bucket, then it gets merged in.
+    let bucket = {
+      key: '',
+      label: '',
+      datasets: [d],
+      datasetIds: [d.dataSetId],
+      dataUse: d.dataUse,
+      dataUses: [],
+      elections: [],
+      votes: [],
+      matchResults: []
+    };
+
+    if (isUndefined(d.dataUse) || isOther(d.dataUse)) {
+      buckets.push(bucket);
+    } else {
+      // Add to bucket if the data use doesn't exist
+      const bucketDataUses = map(b => b.dataUse)(buckets);
+      if (!any(d.dataUse)(bucketDataUses)) {
+        buckets.push(bucket);
+      } else {
+        // If it does exist, merge this dataset into that bucket
+        forEach(b => {
+          if (isEqual(b.dataUse)(d.dataUse)) {
+            b.datasets.push(d);
+            b.datasetIds.push(d.dataSetId);
+          }
+        })(buckets);
+      }
+    }
+
+    // Step 1.a: Find match results for this dataset and add to bucket
+    forEach(m => {
+      if (toLower(d.datasetIdentifier) === toLower(m.consent)) {
+        bucket.matchResults.push(m);
+      }
+    })(matchData);
+
+    // Step 1.b: Populate translated dataUses
+    if (d.dataUse) {
+      const index = findIndex(d.dataUse)(rawDataUses);
+      if (index >= 0 && !isUndefined(flatTranslatedDataUses[index])) {
+        bucket.dataUses = flatTranslatedDataUses[index];
+      }
+    }
+
+  })(datasets);
+
+  // Step 2: Populate elections for each bucket
+  const darMap = get('dars')(collection);
+  forEach(dar => {
+    const electionMap = get('elections')(dar);
+    forEach(election => {
+      // find the bucket this election belongs to
+      const datasetId = get('dataSetId')(election);
+      forEach(b => {
+        if (includes(datasetId)(b.datasetIds)) {
+          b.elections.push(election);
+        }
+      })(buckets);
+    })(values(electionMap));
+  })(values(darMap));
+
+  // Steps 3-5
+  forEach(b => {
+    // Step 3: Generate bucket key and label
+    if (!isEmpty(b.dataUses)) {
+      b.label = flow(
+        map((du) => du.alternateLabel || du.code),
+        join(', ')
+      )(b.dataUses);
+    } else {
+      b.label = 'Undefined Data Use';
+    }
+    b.key = 'bucket-' + join('-')(b.datasetIds);
+
+    // Step 4: Coalesce match results into a single result per bucket
+    b.algorithmResult = calculateAlgorithmResultForBucket(b);
+
+    // Step 5: Populate votes for each bucket
+    b.votes.push(processVotesForBucket(b.elections));
+  })(buckets);
+
+  // Step 6: Populate RUS Vote bucket with RP votes
+  const rpVotes = createRpVoteStructureFromBuckets(buckets);
+  buckets.unshift({
+    isRP: true,
+    key: 'RUS Vote',
+    votes: rpVotes
+  });
```

Comment:
suggestion: structurally, i find the `isRP` confusing, since multiple types are intermixed in one array. what if the returned object was of the structure `{ rpBucket: {}, dataUseBuckets: [{}, {}] }`? i am ok keeping it the same way if it's preferred.

Reply:
TBH, I found it confusing too. However, that's the field that the original version defined and is used in several other functions/components. My goal here was to make as much of a drop-in replacement as possible.